src/chains/algorand_1/algoChainState.ts

```diff
@@ -0,0 +1,187 @@
+
+export class AlgorandChainState {
+  private _activeEndpoint: ChainEndpoint
+
+  private _chainInfo: ChainInfo
+
+  private _chainSettings: AlgorandChainSettings
+
+  private _endpoints: ChainEndpoint[]
+
+  private _isConnected: boolean = false
+
+  private _algo: Algo
+
+  constructor(endpoints: ChainEndpoint[], settings?: AlgorandChainSettings) {
+    this._endpoints = endpoints
+    this._chainSettings = settings
+  }
+
+  /** Return chain URL endpoints */
+  public get activeEndpoint(): ChainEndpoint {
+    return this._activeEndpoint
+  }
+
+  /** * Return chain ID */
+  public get chainId(): string {
+    this.assertIsConnected()
+    return this._chainInfo?.nativeInfo?.chain_id
+  }
+
+  /** Return chain info - e.g. head block number */
+  public get chainInfo(): ChainInfo {
+    this.assertIsConnected()
+    return this._chainInfo
+  }
+
+  /** Return chain settings */
+  public get chainSettings(): AlgorandChainSettings {
+    return this._chainSettings
+  }
+
+  /** Return chain URL endpoints */
+  public get endpoints(): ChainEndpoint[] {
+    return this._endpoints
+  }
+
+  /** Whether any information has been retrieved from the chain yet
+   *   A value here confirms that the network is working
+   * */
+  public get isConnected(): boolean {
+    return this._isConnected
+  }
+
+  /** Connect to chain endpoint to verify that it is operational and to get latest block info */
+  public async connect(): Promise<void> {
+    try {
+      if (!this._algo) {
+        const { token, server, port } = this.determineAlgorandConnectionSettings()
+        this._algo = new algosdk.Algod(token, server, port)
+      }
+      await this.getChainInfo()
+      this._isConnected = true
+    } catch (error) {
+      throwAndLogError('Problem connecting to chain', 'chainConnectFailed', error)
+    }
+  }
+
+  /** Retrieve lastest chain info including head block number and time */
+  public async getChainInfo(): Promise<ChainInfo> {
+    try {
+      const nodeInfo = await this._algo.status()
+      const { lastRound, lastConsensusVersion, timeSinceLastRound } = nodeInfo
+      this._chainInfo = {
+        headBlockNumber: lastRound,
+        headBlockTime: new Date(timeSinceLastRound),
+        // version example: 'https://github.com/algorandfoundation/specs/tree/e5f565421d720c6f75cdd186f7098495caf9101f'
+        version: lastConsensusVersion,
+        nativeInfo: nodeInfo,
+      }
+      return this._chainInfo
+    } catch (error) {
+      // ALGO TODO: map chain error
+      throw new Error()
+    }
+  }
+
+  // TODO: sort based on health info
+  /**  * Choose the best Chain endpoint based on health and response time */
+  private determineAlgorandConnectionSettings(): AlgorandConnectionSettings {
+    const urlSetting = this.endpoints[0]
+    const url = trimTrailingChars(urlSetting.url.href, '/')
+    const { token, port = '' } = urlSetting.settings
+    return { server: url, token, port }
```

Comment:
Pull token from the headers in endpoint settings once the branch is rebased against master.